utility/src/main/java/org/oppia/util/parser/LiTagHandler.kt

```diff
@@ -1,37 +1,171 @@
 package org.oppia.util.parser
 
+import android.content.Context
 import android.text.Editable
 import android.text.Html
 import android.text.Spannable
 import android.text.Spanned
-import android.text.style.BulletSpan
+import android.text.Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
+import android.text.Spanned.SPAN_MARK_MARK
+import android.text.style.LeadingMarginSpan
+import java.util.Stack
+import org.oppia.util.parser.LiTagHandler.ListTag
+import org.oppia.util.parser.StringUtils.LI_TAG
+import org.oppia.util.parser.StringUtils.OL_TAG
+import org.oppia.util.parser.StringUtils.UL_TAG
 import org.xml.sax.XMLReader
 
 /**
- * [Html.TagHandler] implementation that processes <li> tags and creates bullets.
+ * Called when the HTML parser reaches an opening or closing tag.
+ * We only handle list tags and ignore other tags.
  *
- * Reference: https://github.com/davidbilik/bullet-span-sample
+ * <ul> and <ol> tags are pushed to the [lists] stack and popped when the closing tag is reached.
+ *
+ * <li> tags are handled by the [ListTag] instance corresponding to the parent tag.
+ *
+ * Reference: https://github.com/daphliu/android-spannable-list-sample/tree/master/app/src/main/java/com/daphneliu/sample/listspansample
  */
-class LiTagHandler : Html.TagHandler {
+class LiTagHandler(private val context: Context) : Html.TagHandler {
+
+  private val lists = Stack<ListTag>()
+
+  override fun handleTag(opening: Boolean, tag: String, output: Editable, xmlReader: XMLReader) {
+    when (tag) {
+      UL_TAG -> if (opening) {
+        // handle <ul>
+        lists.push(Ul())
+      } else {
+        // handle </ul>
+        lists.pop()
+      }
+      OL_TAG -> if (opening) {
+        // handle <ol>
+        lists.push(Ol())
+      } else {
+        // handle </ol>
+        lists.pop()
+      }
+      LI_TAG -> if (opening) {
+        // handle <li>
+        lists.peek().openItem(output)
+      } else {
+        // handle </li>
+        lists.peek().closeItem(output, indentation = lists.size - 1)
+      }
+    }
+  }
+
   /**
-   * Helper marker class. Based on [Html.fromHtml] implementation.
+   * Handler for <li> tags. Subclasses set the bullet appearance.
    */
-  class Bullet
+  private interface ListTag {
```

Comment:
I think this might be a good place to leverage a sealed class. See: https://antonioleiva.com/sealed-classes-kotlin/ and https://kotlinlang.org/docs/reference/sealed-classes.html.

Reply:
I am not so familiar with this. I will study this and then implement it.